### PR TITLE
relative path support at providers-url of repo and dist url of package

### PR DESCRIPTION
**Pull request by mipxtx**
I need to access the local repo through a many urls. 
This will add a feature of relative urls with base, provided by client ( "repositories": [{"url":<url>}... at  composer.json)

### CODE REVIEW

File: src/Composer/Repository/ComposerRepository.php [hunk 1]
```diff
@@ -526,9 +526,11 @@ protected function canonicalizeUrl($url)
     {
         if ('/' === $url[0]) {
             return preg_replace('{(https?://[^/]+).*}i', '$1' . $url, $this->url);
+        } elseif (strpos($url, "http://") === 0 || strpos($url, "https://") === 0) {
+            return $url;
         }
 
-        return $url;
+        return $this->url . "/" . $url;
```
Comment (stof) [inline src/Composer/Repository/ComposerRepository.php]: this is broken in case the URL is ``http://packagist.org/packages.json``, which is perfectly supported.

Comment (mipxtx) [inline src/Composer/Repository/ComposerRepository.php]: line 529: elseif (strpos($url, "http://") === 0 || strpos($url, "https://") === 0)  filters this link. 
can you get me a composer.json description,  to reproduce a bug. 
My test was: 
composer.json: 
{
  "require": {
    "psr/log": "*"
  }
} 

composer debug output: 
 ../composer/bin/composer -vvv update 
Reading ./composer.json
Loading config file ./composer.json
Checked CA file /opt/local/share/curl/curl-ca-bundle.crt: valid
Executing command (/Users/mix/wsp/test): git branch --no-color --no-abbrev -v
Executing command (/Users/mix/wsp/test): git describe --exact-match --tags
Executing command (/Users/mix/wsp/test): git log --pretty="%H" -n1 HEAD
Failed to initialize global composer: Composer could not find the config file: /Users/mix/.composer/composer.json
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
Running @package_version@ (@release_date@) with PHP 7.0.5 on Darwin / 15.5.0
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Loading composer repositories with package information
Downloading https://packagist.org/packages.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/packages.json into cache
Updating dependencies (including require-dev)
Downloading http://packagist.org/p/provider-2013%24d841c1c8e6a0a08dee090ac9fd44fac96e8bf072ec70151504e7a9a36f35b9e3.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-2013.json into cache
Downloading http://packagist.org/p/provider-2014%24367ce7ccd1e98fdd934e5567aed3cb5e9081a979222ce86f98b754e3ee619d9b.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-2014.json into cache
Downloading http://packagist.org/p/provider-2015%24e6584783bb345dba8830169f9ee93cb1c22a20d4d621bfb7470b6d548f65a934.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-2015.json into cache
Downloading http://packagist.org/p/provider-2015-10%247769e33e9d6ec8e5ba839eadc2c02a6293ae95c70f62dd1cd69e85959c4dded8.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-2015-10.json into cache
Downloading http://packagist.org/p/provider-2016-01%249ab22e7ac67f30d8d5cb62a439ef6f8d91ae6bfc2bb0d00aabf5e8c312116fa8.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-2016-01.json into cache
Downloading http://packagist.org/p/provider-2016-04%2453399185e484493d7240221db431388b88d88ab9f3808b07cee5f791e71bb07e.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-2016-04.json into cache
Downloading http://packagist.org/p/provider-2016-07%24045abadb19adba9a8964490d4c7d733c619f3253360fda46a2bc04139acd392a.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-2016-07.json into cache
Downloading http://packagist.org/p/provider-archived%24c4dd02a51d350c40e1a81d8b754638deebedd8a69b37ba10813c05495de8fe01.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-archived.json into cache
Downloading http://packagist.org/p/provider-latest%24fbd2d9f6254a64e87a14dafb1c9c13a15f9a42dedeb6d39c65eb38bc844cd294.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/p-provider-latest.json into cache
Downloading http://packagist.org/p/psr/log%24d57098a03fb1141c0e8a3915b0561677f95ec6d7534e2b7a95ed257119dbf1be.json
Writing /Users/mix/.composer/cache/repo/https---packagist.org/provider-psr$log.json into cache
Resolving dependencies through SAT
Dependency resolution completed in 0.001 seconds
Analyzed 54 packages to resolve dependencies
Analyzed 54 rules to resolve dependencies
  - Installing psr/log (1.0.0)
Downloading https://api.github.com/repos/php-fig/log/zipball/fe0936ee26643249e916849d48e3a51d5f5e278b
    Downloading: 100%         
Writing /Users/mix/.composer/cache/files/psr/log/566a6d690c0340b7292a7eb95b91042cf8524c40.zip into cache from /Users/mix/wsp/test/vendor/psr/log/48a053673a86a8836fa452b66d8b85c2
    Extracting archive
Executing command (CWD): unzip '/Users/mix/wsp/test/vendor/psr/log/48a053673a86a8836fa452b66d8b85c2' -d '/Users/mix/wsp/test/vendor/composer/7c8a5677' && chmod -R u+w '/Users/mix/wsp/test/vendor/composer/7c8a5677'

    REASON: Required by the root package: Install command rule (install psr/log 1.0.0)

Reading /Users/mix/wsp/test/vendor/composer/installed.json
Reading ./composer.lock
Writing lock file
Generating autoload files

My output shows a correct fetch of https://packagist.org/packages.json

Comment (stof) [inline src/Composer/Repository/ComposerRepository.php]: I'm talking about the URL of the repo, not about the URL of the provider

File: tests/Composer/Test/Downloader/XzDownloaderTest.php [hunk 2]
```diff
@@ -72,7 +72,12 @@ public function testErrorMessages()
             $downloader->download($packageMock, $this->getUniqueTmpDirectory());
             $this->fail('Download of invalid tarball should throw an exception');
         } catch (\RuntimeException $e) {
-            $this->assertContains('File format not recognized', $e->getMessage());
+            // Mac OS tar has own errors
```
Comment (slbmeh) [inline tests/Composer/Test/Downloader/XzDownloaderTest.php]: That's BSD tar not just Mac... This will fail on Mac that uses GNU tar.

File: src/Composer/Repository/ComposerRepository.php [hunk 3]
```diff
@@ -88,9 +88,16 @@ public function __construct(array $repoConfig, IOInterface $io, Config $config,
...
+            $urlParts = parse_url($this->url);
+            $baseUrlLength = mb_strlen($this->url) - mb_strlen(basename($urlParts['path'])) - 1;
+            $this->baseUrl = mb_strcut($this->url, 0, $baseUrlLength);
+        } else {
+            $this->baseUrl = rtrim($this->url, "/");
+        }
+
         $this->io = $io;
-        $this->cache = new Cache($io, $config->get('cache-repo-dir').'/'.preg_replace('{[^a-z0-9.]}i', '-', $this->url), 'a-z0-9.$');
+        $this->cache = $cache ?: new Cache($io, $config->get('cache-repo-dir').'/'.preg_replace('{[^a-z0-9.]}i', '-', $this->baseUrl), 'a-z0-9.$');
```
Comment (slbmeh) [inline src/Composer/Repository/ComposerRepository.php]: Why does the cache implementation need to change to support a relative url?

File: src/Composer/Repository/ComposerRepository.php [hunk 4]
```diff
@@ -522,13 +527,23 @@ protected function loadRootServerFile()
         return $this->rootData = $data;
     }
 
+    protected function urlIsJsonFile($url)
+    {
+        $jsonUrlParts = parse_url($url);
+
+        return isset($jsonUrlParts['path'])
+            && (mb_strlen($jsonUrlParts['path']) - 5) === strpos($jsonUrlParts['path'], '.json');
```
Comment (slbmeh) [inline src/Composer/Repository/ComposerRepository.php]: Not objecting to the change for ensuring the `.json` is positively at the end, but this is absolutely a BC break.

File: src/Composer/Repository/ComposerRepository.php [hunk 5]
```diff
@@ -522,13 +527,23 @@ protected function loadRootServerFile()
...
+    }
+
     protected function canonicalizeUrl($url)
     {
         if ('/' === $url[0]) {
             return preg_replace('{(https?://[^/]+).*}i', '$1' . $url, $this->url);
+        } elseif (strpos($url, "http://") === 0 || strpos($url, "https://") === 0) {
```
Comment (slbmeh) [inline src/Composer/Repository/ComposerRepository.php]: `preg_match` is faster than the two `strpos` calls here...

```
<?php

$test = range(1000, 2000);

$testdata = [];

$schemes = ['file', 'http', 'https', '/'];

foreach ($schemes as $scheme) {
    $testdata[$scheme] = [];

    foreach ($test as $v) {
        $prefix = ($scheme != '/') ? $scheme . '://' : $scheme;
        $testdata[$scheme][] = $prefix . $v;
    }
}

foreach ($schemes as $scheme) {
    $start = microtime(true);

    foreach ($testdata[$scheme] as $v) {
        $foo = (strpos($v, 'http://' === 0) || strpos($v, 'https://') === 0);
    }

    $stop = microtime(true) - $start;

    echo $scheme . ' strpos execution time: ' . $stop . PHP_EOL;
}

foreach ($schemes as $scheme) {
    $start = microtime(true);

    foreach ($testdata[$scheme] as $v) {
        $foo = (0 !== preg_match('{^(?>https?://)}', $v));
    }

    $stop = microtime(true) - $start;

    echo $scheme . ' preg_match execution time: ' . $stop . PHP_EOL;
}
```

results in:

```
file strpos execution time: 0.001471996307373
http strpos execution time: 0.0014429092407227
https strpos execution time: 0.0014641284942627
/ strpos execution time: 0.0014297962188721
file preg_match execution time: 0.00089597702026367
http preg_match execution time: 0.00078582763671875
https preg_match execution time: 0.00079202651977539
/ preg_match execution time: 0.00081706047058105
```

File: src/Composer/Repository/ComposerRepository.php [hunk 6]
```diff
@@ -88,9 +88,16 @@ public function __construct(array $repoConfig, IOInterface $io, Config $config,
         $this->config = $config;
         $this->options = $repoConfig['options'];
         $this->url = $repoConfig['url'];
-        $this->baseUrl = rtrim(preg_replace('{^(.*)(?:/[^/\\]+.json)?(?:[?#].*)?$}', '$1', $this->url), '/');
+        if ($this->urlIsJsonFile($this->url)) {
+            $urlParts = parse_url($this->url);
+            $baseUrlLength = mb_strlen($this->url) - mb_strlen(basename($urlParts['path'])) - 1;
```
Comment (slbmeh) [inline src/Composer/Repository/ComposerRepository.php]: If the path ends in json, but has a query string or a fragment, this is broken.

File: src/Composer/Repository/ComposerRepository.php [hunk 7]
```diff
@@ -88,9 +88,16 @@ public function __construct(array $repoConfig, IOInterface $io, Config $config,
         $this->config = $config;
         $this->options = $repoConfig['options'];
         $this->url = $repoConfig['url'];
-        $this->baseUrl = rtrim(preg_replace('{^(.*)(?:/[^/\\]+.json)?(?:[?#].*)?$}', '$1', $this->url), '/');
```
Comment (slbmeh) [inline src/Composer/Repository/ComposerRepository.php]: `http://example.com/index.php?page=my/composer/repo/foo.json` will no longer work relying solely on path.

File: tests/Composer/Test/Repository/ComposerRepositoryIncludeTest.php [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+
+namespace Composer\Repository;
+
+use Composer\Cache;
+use Composer\IO\NullIO;
+use Composer\Test\Mock\FactoryMock;
+use Composer\Test\Mock\RemoteFilesystemMock;
+use Composer\TestCase;
+
+class ComposerRepositoryIncludeTest extends TestCase
```
Comment (slbmeh) [inline tests/Composer/Test/Repository/ComposerRepositoryIncludeTest.php]: This test class is extremely hard to follow with the tests having a data provider delegating to a separate function

File: tests/Composer/Test/Repository/ComposerRepositoryIncludeTest.php [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+            array($domain . 'base/repo.json' => json_encode(array('packages' => $package))),
+            $url
+        );
+
+        $package = $packageDist;
+        $package[$name]['1.0.0']['dist']['url']  =  $this->baseUrl;
+
+        #8
+        $out[] = array(
+            array('url' => $this->domain),
```
Comment (slbmeh) [inline tests/Composer/Test/Repository/ComposerRepositoryIncludeTest.php]: Why does this one get `$this->domain` and everything else `$domain` even though `$domain = $this->domain` above.

File: tests/Composer/Test/Repository/ComposerRepositoryIncludeTest.php [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+
+
+        $domain = $this->domain;
+        $package = $this->packageDist;
+        $url = $this->url;
+        $name = $this->packageName;
+
+        $root = array(
+            "providers-url" =>"%package%.json",
+            "providers" => array($this->packageName => array('sha256' => ""))
```
Comment (slbmeh) [inline tests/Composer/Test/Repository/ComposerRepositoryIncludeTest.php]: Why use `$this->packageName` when you use `$name` in the rest of the method?

**Comment by Seldaek** [comment]
If you'd like this to be even considered for merging, we need to know more about the use case. I am not going to merge so much code unless I fully understand what you are trying to do, and especially why you want to do it this way.

**Comment by mipxtx** [comment]
The main idea of this feature is to get portable repository that can be moved between locations without regenerating of json descriptions.
I try to explain this use case. We use composer for deployment of our private packages. And we have two different networks - production and development. Due to security and stability issues developers has no access to production network and production frontends has no access to the development network. Thats why we use two composer repositories. The production repository located at the root of the web server, but the developers repository located at the continuous integration server in subdirectory because the main page contains overview info. Now i should generate different json descriptions of repository for development and production instances, but with this feature i cat just rsync development repo to production. 
The most of my changes(except tests) is replacement of custom handling paths with single resolver. It also fixes a funny bug: you can not specify http://packagist.org/packages.json. it has bug in resolving of provider-includes:

> cat composer.json
> {
>   "repositories": [
>     {"type": "composer","url": "https://packagist.org/packages.json"},
>     {"packagist": false}
>   ],
>   "require": {"symfony/http-foundation": "2.5.4"}
> } 
> ./composer.phar update
> Loading composer repositories with package information
> Updating dependencies (including require-dev)
> 
>                                                                                                                                                                          
>   [Composer\Downloader\TransportException]                                                                                                                               
>   The "https://packagist.org/packages.json/p/provider-2013%24eaa1e86b45c63fa1758f02e985c7d54b2140ae328c0a529a289a81f661b7fe43.json" file could not be downloaded (HTTP/1.1 404 Not Found)

**Comment by Seldaek** [comment]
Thanks for the explanation, but I am still not sure I fully understand, sorry. Can you provide an example composer.json of how you use it? Do you have then a different composer repository defined for prod and dev? How/why are relative URLs required? Why can't the two different repos use different absolute paths and that's it? Or could you not use DNS overrides so that prod servers for example resolve your repo's IP to the prod repo's IP instead of the dev one?

As for specifying a URL with /package.json at the end, this is fixed in master and will be fixed in the next release.

**Comment by mipxtx** [comment]
We use two different composer repositories for prod and dev. And we not use DNS overrides and other network hacks. In this case we use two different composer.json at dev and prod. It looks like 
{
  "repositories": [
    {
      "type": "composer",
      "url": "http://packagist.lan/"
    },
    {"packagist": false}
  ],
  "config": { "secure-http": false},
  "require": {"mamba/mamba":"*"}
}
for production and 
{
  "repositories": [
    {
      "type": "composer",
      "url": "http://deploy.lan/repo"
    },
    {"packagist": false}
  ],
  "config": { "secure-http": false},
  "require": {"mamba/mamba":"*"}
} 
at dev. 
They differs with the url of composer repo. 

deploy.lan server is the complex system of ci, git repositories, some testing stuff. Also it contains a composer repository at the subfolder of the root of web server. 
packagist.lan is a separate production repo server. Repository located at the root of the web server. This is the main problem: i must generate different packages.json file and includes both for dev and prod servers. 
prod packages.json looks like:
{
    "providers-url": "\/%package%\/package-%hash%.json",
    "providers": {
        "phpunit\/phpunit": {
            "sha256": "123123"
        },
    ...
    }
}
and dev packages.json looks like 
{
    "providers-url": "\/repo\/%package%\/package-%hash%.json",
    "providers": {
        "phpunit\/phpunit": {
            "sha256": "123123"
        },
    ...
    }
}
json files of packages is also differs, because it contains links to package versions that begins with "\/repo" at dev and "\/" at prod.

Relative links helps me to get mirror of dev repository at production network.
packages.json will be looks like 
{
    "providers-url": "%package%\/package-%hash%.json",
    "providers": {
        "phpunit\/phpunit": {
            "sha256": "123123"
        },
    ...
    }
}

in this case all files can be downloaded with the base, that described in the composer.json of client.  "http://deploy.lan/repo/" + "%package%\/package-%hash%.json" at the dev network and "http://packagist.lan/" + "%package%\/package-%hash%.json" at production.

**Comment by Seldaek** [comment]
OK I see.. but would it not be easier to move it as well to /repo on the production server than to add this into composer? :)

**Comment by mipxtx** [comment]
Yes, i can store my repo at the same folder, but it doesn't "right". This is unnecessary filesystem dependency, that can be risky at server management. You can move git repository and use it without rebuilding. Yum also supports location change at my centos. I suppose this feature has being helpful for society. If its not, you can close this pull request.
